Sources/APIServer/Volumes/VolumesService.swift

```diff
@@ -0,0 +1,220 @@
+        let fm = FileManager.default
+        try fm.createDirectory(atPath: volumePath, withIntermediateDirectories: true, attributes: nil)
+    }
+
+    private func createVolumeImage(for name: String, sizeInBytes: UInt64 = VolumeStorage.defaultVolumeSizeBytes) throws {
+        let blockPath = blockPath(for: name)
+
+        // Use the containerization library's EXT4 formatter
+        let formatter = try EXT4.Formatter(
+            FilePath(blockPath),
+            blockSize: 4096,
+            minDiskSize: sizeInBytes
+        )
+
+        try formatter.close()
+    }
+
+    private func removeVolumeDirectory(for name: String) throws {
+        let volumePath = volumePath(for: name)
+        let fm = FileManager.default
+
+        if fm.fileExists(atPath: volumePath) {
+            try fm.removeItem(atPath: volumePath)
+        }
+    }
+
+    private func _create(
+        name: String,
+        driver: String,
+        driverOpts: [String: String],
+        labels: [String: String]
+    ) async throws -> Volume {
+        guard VolumeStorage.isValidVolumeName(name) else {
+            throw VolumeError.invalidVolumeName("Invalid volume name '\(name)': must match \(VolumeStorage.volumeNamePattern)")
+        }
+
+        // Check if volume already exists by trying to list and finding it
+        let existingVolumes = try await store.list()
+        if existingVolumes.contains(where: { $0.name == name }) {
+            throw VolumeError.volumeAlreadyExists(name)
+        }
+
+        try createVolumeDirectory(for: name)
+
+        // Parse size from driver options (default 512GB)
+        let sizeInBytes: UInt64
+        if let sizeString = driverOpts["size"] {
+            sizeInBytes = try parseSize(sizeString)
+        } else {
+            sizeInBytes = VolumeStorage.defaultVolumeSizeBytes
+        }
+
+        try createVolumeImage(for: name, sizeInBytes: sizeInBytes)
+
+        let volume = Volume(
+            name: name,
+            driver: driver,
+            format: "ext4",
+            source: blockPath(for: name),
+            labels: labels,
+            options: driverOpts
+        )
+
+        try await store.create(volume)
+
+        log.info("Created volume", metadata: ["name": "\(name)", "driver": "\(driver)"])
+        return volume
+    }
+
+    private func _delete(name: String) async throws {
+        guard VolumeStorage.isValidVolumeName(name) else {
+            throw VolumeError.invalidVolumeName("Invalid volume name '\(name)': must match \(VolumeStorage.volumeNamePattern)")
+        }
+
+        // Check if volume exists by trying to list and finding it
+        let existingVolumes = try await store.list()
+        guard existingVolumes.contains(where: { $0.name == name }) else {
+            throw VolumeError.volumeNotFound(name)
+        }
+
+        // Check if volume is in use by any container atomically
+        let volumeInUse = try await containersService.withContainerList { containers in
```

Comment:
Shouldn't the volume delete itself be executed in the block so that there's no chance of a reference being created to the volume after this block exits? Right now it seems that if another thread is creating a container that attaches the volume at the same time we're at line 197, there's a race still.